### PR TITLE
Onboard EventHubs onto structured logging

### PR DESCRIPTION
**Pull request by lmolkova**
Context: https://gist.github.com/lmolkova/04f6484595fdd7edded9f4d82619291f

Previous PRs:
- #25247 - core changes
- #25671 - azure-core-amqp changes

### CODE REVIEW

File: sdk/eventhubs/azure-messaging-eventhubs-checkpointstore-blob/src/main/java/com/azure/messaging/eventhubs/checkpointstore/blob/BlobCheckpointStore.java [hunk 1]
```diff
@@ -50,6 +50,13 @@
     private static final String CHECKPOINT_PATH = "/checkpoint/";
     private static final String OWNERSHIP_PATH = "/ownership/";
 
+    // logging keys, consistent across all AMQP libraries and human-readable
+    private static final String PARTITION_ID_LOG_KEY = "partitionId";
+    private static final String OWNER_ID_LOG_KEY = "ownerId";
+    private static final String SEQUENCE_NUMBER_LOG_KEY = "sequenceNumber";
+    private static final String BLOB_NAME_LOG_KEY = "sequenceNumber";
```
Comment (JonathanGiles) [inline sdk/eventhubs/azure-messaging-eventhubs-checkpointstore-blob/src/main/java/com/azure/messaging/eventhubs/checkpointstore/blob/BlobCheckpointStore.java:57]: This should be `blobName`

**Comment by ramya-rao-a** [comment]
Hey @lmolkova!

I am interested in understanding more about how we are doing structured logging. Could you perhaps consider updating the PR description with some context?

**Review by conniey — APPROVED** [review]
LGTM!

File: sdk/eventhubs/azure-messaging-eventhubs/src/samples/java/com/azure/messaging/eventhubs/SampleCheckpointStore.java [hunk 2]
```diff
@@ -81,9 +84,11 @@ private String prefixBuilder(String fullyQualifiedNamespace, String eventHubName
                     || partitionOwnershipMap.get(partitionOwnership.getPartitionId()).getETag()
                     .equals(partitionOwnership.getETag());
             })
-            .doOnNext(partitionOwnership -> logger
-                .info("Ownership of partition {} claimed by {}", partitionOwnership.getPartitionId(),
-                    partitionOwnership.getOwnerId()))
+            .doOnNext(partitionOwnership ->
+                logger.atInfo()
+                    .addKeyValue(PARTITION_ID_KEY, partitionOwnership.getPartitionId())
+                    .addKeyValue("ownerId", partitionOwnership.getOwnerId())
```
Comment (conniey) [inline sdk/eventhubs/azure-messaging-eventhubs/src/samples/java/com/azure/messaging/eventhubs/SampleCheckpointStore.java:90]: Is there a reason we don't use the constant here?

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/resources/eventhubs-messages.properties [hunk 3]
```diff
@@ -4,11 +4,11 @@ REQUEST_VALUE_NOT_VALID=Back pressure request value not valid. It must be betwee
 EVENT_DATA_DOES_NOT_FIT=EventData does not fit into maximum number of batches. '%s'
 CANNOT_SEND_EVENT_BATCH_EMPTY=Cannot send an EventBatch that is empty.
 ERROR_SENDING_BATCH=Error sending batch.
-FAILED_TO_CLAIM_OWNERSHIP=Failed to claim ownership of partition {} - {}
-LOAD_BALANCING_FAILED=Load balancing for event processor failed - {}
+FAILED_TO_CLAIM_OWNERSHIP=Failed to claim ownership
+LOAD_BALANCING_FAILED=Load balancing for event processor failed.
```
Comment (conniey) [inline sdk/eventhubs/azure-messaging-eventhubs/src/main/resources/eventhubs-messages.properties:8]: Nit since we added a period here that we could add them to the rest of the sentences.

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/SynchronousEventSubscriber.java [hunk 4]
```diff
@@ -87,18 +97,17 @@ public void dispose() {
...
-            this.workId = workId;
+        ReceiveTimeoutTask(Runnable onDispose, ClientLogger logger) {
             this.onDispose = onDispose;
+            this.logger = logger;
         }
 
         @Override
         public void run() {
-            logger.info("Work: {}. Timeout encountered, disposing of subscriber.", workId);
+            logger.info("Timeout encountered, disposing of subscriber.");
```
Comment (conniey) [inline sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/SynchronousEventSubscriber.java:110]: Do we not need the workId anymore or to know which one timed out?

**Comment by lmolkova** [comment]
> I am interested in understanding more about how we are doing structured logging. Could you perhaps consider updating the PR description with some context?

@ramya-rao-a sure, I updated the description. 

**TL;DR:**

We write a portion of the log message that SDK controls as json. Logging APIs now allow adding key-value-pairs. As a result, we're switching from 

```2021-10-01 06:32:42,732 INFO [com.azure.core.amqp.implementation.handler.ReceiveLinkHandler] (reactor-executor-183) connectionId[MF_c991bc_1632981438863] linkName[session-_368299_1633069903838] entityPath[pms-restrictions] Receiver link was never active. Closing endpoint states.```

to 

```2021-10-01 06:32:42,732 INFO [com.azure.core.amqp.implementation.handler.ReceiveLinkHandler] (reactor-executor-183) {"az.sdk.message": "Receiver link was never active. Closing endpoint states", "connectionId":"MF_c991bc_1632981438863", "linkName":"session-_368299_1633069903838", "entityPath":"pms-restrictions"}```

These logs (when shared as files) can be easily parsed with [some tooling](https://github.com/lmolkova/azure-sdk-log-parser), sent to Log Analytics and queried. 
They could also be analyzed by customers: e.g. Azure Monitor allows to parse json at query time. Splunk and some other logging tools also allow parsing json at query time.

I.e. we can semi-automate log analysis, or give queries to support engineers and customers to troubleshoot common issues.

File: sdk/eventhubs/azure-messaging-eventhubs/src/samples/java/com/azure/messaging/eventhubs/SampleCheckpointStore.java (hunk 2 above)
Comment (lmolkova) [reply sdk/eventhubs/azure-messaging-eventhubs/src/samples/java/com/azure/messaging/eventhubs/SampleCheckpointStore.java]: great catch!
